core/src/main/scala/org/apache/spark/api/python/VirtualEnvFactory.scala

```diff
@@ -0,0 +1,164 @@
+      s"VirtualEnvBinPath: $virtualEnvBinPath is not defined or doesn't exist.")
+    // Two scenarios of creating virtualenv:
+    // 1. created in yarn container. Yarn will clean it up after container is exited
+    // 2. created outside yarn container. Spark need to create temp directory and clean it after app
+    //    finish.
+    //      - driver of PySpark shell
+    //      - driver of yarn-client mode
+    if (isLauncher ||
+      (isDriver && conf.get("spark.submit.deployMode") == "client")) {
+      val virtualenvBasedir = Files.createTempDir()
+      virtualenvBasedir.deleteOnExit()
+      virtualEnvName = virtualenvBasedir.getAbsolutePath
+    } else if (isDriver && conf.get("spark.submit.deployMode") == "cluster") {
+      virtualEnvName = "virtualenv_driver"
+    } else {
+      // use the working directory of Executor
+      virtualEnvName = "virtualenv_" + conf.getAppId + "_" + VIRTUALENV_ID.getAndIncrement()
+    }
+
+    // Use the absolute path of requirement file in the following cases
+    // 1. driver of pyspark shell
+    // 2. driver of yarn-client mode
+    // otherwise just use filename as it would be downloaded to the working directory of Executor
+    val pysparkRequirements =
+      if (isLauncher ||
+        (isDriver && conf.get("spark.submit.deployMode") == "client")) {
+        conf.getOption("spark.pyspark.virtualenv.requirements")
+      } else {
+        conf.getOption("spark.pyspark.virtualenv.requirements").map(_.split("/").last)
+      }
+
+    val createEnvCommand =
+      if (virtualEnvType == "native") {
+        List(virtualEnvBinPath,
+          "-p", pythonExec,
+          "--no-site-packages", virtualEnvName)
+      } else {
+        // Two cases of conda
+        //    1. requirement file is specified. (Batch mode)
+        //    2. requirement file is not specified. (Interactive mode).
+        //       In this case `spark.pyspark.virtualenv.python_version` must be specified.
+
+        if (pysparkRequirements.isDefined) {
+          List(virtualEnvBinPath,
+            "create", "--prefix", virtualEnvName,
+            "--file", pysparkRequirements.get, "-y")
+        } else {
+          require(conf.contains("spark.pyspark.virtualenv.python_version"),
+            "spark.pyspark.virtualenv.python_version is not set when using conda " +
+              "in interactive mode")
+          val pythonVersion = conf.get("spark.pyspark.virtualenv.python_version")
+          List(virtualEnvBinPath,
+            "create", "--prefix", virtualEnvName,
+            "python=" + pythonVersion, "-y")
+        }
+      }
+    execCommand(createEnvCommand)
+
+    virtualPythonExec = virtualEnvName + "/bin/python"
+    if (virtualEnvType == "native" && pysparkRequirements.isDefined) {
+      // requirement file for native is not mandatory, run this only when requirement file
+      // is specified.
+      execCommand(List(virtualPythonExec, "-m", "pip",
+        "--cache-dir", System.getProperty("user.home"),
```

Comment:
Why chacing in the user home dir?

Reply:
In yarn mode, executor runs under user `yarn`, while pip would store cache in other directory that `yarn` don't have permission to write. So here I specify the cache dir to yarn's home directory.

Reply:
this builds a cache for each user where you'd end up with multiple caches for the same packages. how about a config to set a common cache (which would have to be writeable by everybody of course)?

also i'm wondering about runaway disk usage. who cleans up when things get tight?